docs/basics/part-7-using-docker-with-earthly.md

```diff
@@ -0,0 +1,541 @@
+To copy the files for [this example ( Part 7 )](https://github.com/earthly/earthly/tree/main/examples/tutorial/go/part7) run
+
+```bash
+earthly --artifact github.com/earthly/earthly/examples/tutorial/go:main+part7/part7 ./part7
+```
+
+Examples in [Python](#more-examples), [Javascript](#more-examples) and [Java](#more-examples) are at the bottom of this page.
+
+## The WITH DOCKER Command
+
+You may find that you need to run Docker commands inside of a target. For those cases Earthly offers `WITH DOCKER`. `WITH DOCKER` will initialize a Docker daemon that can be used in the context of a `RUN` command.
+
+Whenever you need to use `WITH DOCKER` we recommend (though it is not required) that you use Earthly's own Docker in Docker (dind) image: `earthly/dind:alpine`.
+
+Notice `WITH DOCKER` creates a block of code that has an `END` keyword. Everything that happens within this block is going to take place within our `earthly/dind:alpine` container.
+
+### Pulling an Image
+```Dockerfile
+hello:
+    FROM earthly/dind:alpine
+    WITH DOCKER --pull hello-world
+        RUN docker run hello-world
+    END
+
+```
+You can see in the command above that we can pass a flag to `WITH DOCKER` telling it to pull an image from Docker Hub. We can pass other flags to [load in artifacts built by other targets](#loading-an-image) `--load` or even images defined by [docker-compose](#a-real-world-example) `--compose`. These images will be available within the context of `WITH DOCKER`'s docker daemon.
+
+### Loading an Image
+We can load in an image created by another target with the `--load` flag.
+
+```Dockerfile
+my-hello-world:
+    FROM ubuntu
+    CMD echo 'hello world'
+    SAVE IMAGE my-hello:latest
+
+hello:
+    FROM earthly/dind:alpine
+    WITH DOCKER --load hello:latest=+my-hello-world
+        RUN docker run hello
+    END
+```
+
+## A Real World Example
+
+One common use case for `WITH DOCKER` is running integration tests that require other services. In this case we need to set up a redis service for our tests. For this we can user a `docker-compose.yml`.
+
+`docker-compose.yml`
+```yml
+version: "3"
+services:
+  redis:
+    container_name: local-redis
+    image: redis:6.0-alpine
+    ports:
+      - 127.0.0.1:6379:6379
+    hostname: redis
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:6379"]
+      interval: 1s
+      timeout: 10s
+      retries: 5
+
+```
+`main.go`
+```go
+package main
+
+import (
+	"github.com/sirupsen/logrus"
+)
+
+var howCoolIsEarthly = "IceCool"
+
+func main() {
+	logrus.Info("hello world")
+}
+```
+`main_integration_test.go`
+```go
+package main
+
+import (
+	"context"
+	"testing"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIntegration(t *testing.T) {
+	ctx := context.Background()
+	rdb := redis.NewClient(&redis.Options{
+		Addr:     "localhost:6379",
```

Comment:
Ideally this is simply:

```suggestion
		Addr:     "redis:6379",
```

You might need to remove `container_name` and `hostname` from the docker-compose.yml.